### PR TITLE
#162725265 Users should be able to receive notifications

### PR DESCRIPTION
**Pull request by malfahad**
#### What does this PR do?
This PR handles in-app notifications and email notifications.
Notifications give users to opportunity to know what they may have missed which may be important to them.

#### Description of Task to be completed?
There may be 2 types of notifications: email notifications and in-app notifications.

Email notifications are sent as email messages to the email address the user submitted during registration.
Users should be able to opt out (and in) of this type of notification.

In-app notifications show up in the app. This may be visible on a notification page or some other view, as the team may choose.

#### How should this be manually tested?

In-app notifications can be created, read, and marked as read. 

To Test this follow the steps below:
 - Fetch and check into this branch with `git fetch origin ft-notifications-162725265 && git checkout ft-notifications-162725265`
 - setup and activate your virtual environment
 - make migrations with `python manage.py migrate`
 - run the server with `python manage.py runserver`
 - Test the following endpoints with Postman

This PR so far addresses has the following endpoints
`GET /api/notifications/all` 
This fetches all notifications of an authenticated user

`GET /api/notifications/unread`
This fetched unread notifications

`PUT /api/notifications/all`
This marks all notifications as read

##### notification permisions:

`GET /api/notifications/perms<token>` 
 User is opted out of notificaion type encoded in token

`PUT /api/notifications/perms/<notification_type>`
User opts into a specific notification type such as `PROFILE_FOLLOW` and `RECEIVE_NOTIFICATION_EMAILS`

#### What are the relevant pivotal tracker stories?
[162725265](https://www.pivotaltracker.com/story/show/162725265)

#### FUNCTIONALITY TESTING INSTRUCTIONS

1. Login to your author's haven account (use a valid email for this)
2. Fetch all notifications at `GET /api/notifications/all` 
3. The result should be an empty notifications array
4. Create an article and save its slug in your environment variable
5. Log in as a different User and Leave a comment on the created article
6. Log back in as the author and 'GET /api/notifications/unread'
7. You should see your in-app notifications
8. To mark them as read, `PUT /api/notifications/all`
9. Log into the author's email client such as [gmail](mail.google.com) to see the email notification.
10. copy the notification opt-out URL from the email message and paste it in the browser
11. You should see a successful opt-out message
12. To opt back into the notification do `PUT /api/notifications/perms/<notification_type>` as an
authenticated user.

##### POSTMAN
[Run in Postman](https://app.getpostman.com/run-collection/e76f0d7329a979325f26)

### CODE REVIEW

**Comment by ja-odur** [comment]
@malfahad Please rebase and fix the issues on this PR.

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -180,4 +180,3 @@ def put(self, request,token):
         user.set_password(serializer.data.get('password'))
         user.save()
         return Response({"password-reset":"Your password has been updated"}, status=status.HTTP_200_OK)
```
Comment (ja-odur) [inline authors/apps/authentication/views.py]: @malfahad Please avoid unnecessary changes

File: authors/apps/notifications/migrations/0002_auto_20190122_1251.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.1.5 on 2019-01-22 12:51
+
```
Comment (ja-odur) [inline authors/apps/notifications/migrations/0002_auto_20190122_1251.py]: @malfahad please generate a single migration file for a feature

File: authors/apps/profiles/views.py [hunk 3]
```diff
@@ -57,3 +57,4 @@ class UserProfileListView(generics.ListAPIView):
 
     def get_queryset(self):
         return Profile.objects.all()
+
```
Comment (ja-odur) [inline authors/apps/profiles/views.py]: @malfahad avoid unnecessary changes

**Comment by ja-odur** [comment]
@malfahad Please remove the binary files.

File: .travis.yml [hunk 4]
```diff
@@ -3,10 +3,8 @@ python:
     - "3.6.5"
 addons:
   postgresql: "9.6"
-
-services:
+
```
Comment (ja-odur) [inline .travis.yml]: @malfahad undo this change inorder for travis to run your branch

File: authors/apps/notifications/backends.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
+from .models import Notification,NotificationManager
+from django.urls import reverse
+from authors.apps.core.models import email_dispatch
+from .utils import permissions
+import threading
+
```
Comment (armstrongsouljah) [inline authors/apps/notifications/backends.py:6]: missing line here

File: authors/apps/notifications/apps.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+from django.apps import AppConfig
+
+
+class ProfileConfig(AppConfig):
+    name = 'notifications'
```
Comment (armstrongsouljah) [inline authors/apps/notifications/apps.py]: missing line here.

File: authors/apps/core/models.py [hunk 7]
```diff
@@ -76,3 +77,51 @@ def send(self,email_text):
...
+        """Sends email_text to recipient"""
+        return send_mail(
+            self.subject,
+            self.message,
+            self.sender_email,
+            [self.receiver_email],
+            fail_silently=True
+        )
+
+email_dispatch = EmailNotificationDispatch()
```
Comment (armstrongsouljah) [inline authors/apps/core/models.py]: missing line here.

File: authors/apps/core/models.py [hunk 8]
```diff
@@ -76,3 +77,51 @@ def send(self,email_text):
...
+        return jwt.encode({'email': email,
+                            'permission_type':permission_type,
+                             'exp': datetime.datetime.utcnow() + datetime.timedelta(days=30)
+                             },
+                            settings.SECRET_KEY
+                            ,algorithm='HS256')
+
+    def decode_token(self, token):
+        try:
+            payload = jwt.decode(token,settings.SECRET_KEY,algorithm='HS256')
```
Comment (armstrongsouljah) [inline authors/apps/core/models.py]: add spaces after each argument.

File: authors/apps/core/models.py [hunk 9]
```diff
@@ -76,3 +77,51 @@ def send(self,email_text):
...
+            return payload['email'],payload['permission_type']
+        except:
+            raise PermissionError('Invalid or Expired token')
+        
+    def build_and_send(self,request, notification, notification_permission):
+        self.receiver_email = notification.receiver.user.email
+        opt_out_url = request.build_absolute_uri(reverse('notifications:perms', kwargs = {'permission_type':self.make_token(self.receiver_email,notification_permission)}))
```
Comment (armstrongsouljah) [inline authors/apps/core/models.py]: Line exceeds the 86 character limit!

**Comment by mariamiah** [comment]
screen shot 2019-01-28 at 5 54 51 pm

Please change this to git fetch origin

**Comment by ja-odur** [comment]
@malfahad please remake migrations.

**Comment by ja-odur** [comment]
@malfahad according to your description of how to manually test the PR, why do we have to save the slug to the environment variables?

File: authors/apps/comments/migrations/0002_auto_20190124_1741.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.1.5 on 2019-01-24 17:41
+
```
Comment (MuhweziDeo) [inline authors/apps/comments/migrations/0002_auto_20190124_1741.py:2]: @malfahad  thought we only push one migration file

**Comment by mariamiah** [comment]
@malfahad please check code climate

**Review by armstrongsouljah — COMMENTED** [review]
I was able to receive notifications both in-app and from the email. Moving on to the logic.

File: authors/apps/notifications/utils.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    'FAVORITED_ARTICLE_LIKED': 6,
+    'AUTHORED_ARTICLE_COMMENTED': 7,
+    'AUTHORED_ARTICLE_LIKED': 8,
+    'AUTHORED_ARTICLE_FAVORITED': 9
+    } 
+    
+   
+    def get_all(self):
+        return list(self.all.keys())
+
```
Comment (armstrongsouljah) [inline authors/apps/notifications/utils.py]: should be 2 lines separating the class and its instance.

File: authors/apps/notifications/views.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    renderer_classes = (NotificationJSONRenderer,)
+    serializer_class = NotificationSerializer
+
+    def get_permissions(self):
+        self.permission_classes = [AllowAny, ] if self.request.method == 'GET'\
+         else  [IsAuthenticated, ]
+
+        return super(NotiificationsOptOutAPIView, self).get_permissions()
+
+
```
Comment (armstrongsouljah) [inline authors/apps/notifications/views.py]: reduce these to just one line since both functions belong to the same class.

File: authors/apps/notifications/views.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        if permission_type in me.notification_perms:
+            return Response(
+            {'permissions':'You already opted into this notification'},
+                     status=status.HTTP_200_OK)
+        me.notification_perms.append(permission_type)
+        me.save()
+        return Response(
+            {'permissions':'Your notification opt in is successful.'},
+                     status=status.HTTP_200_OK)
+
```
Comment (armstrongsouljah) [inline authors/apps/notifications/views.py]: Add the missing line at the end of this file.

**Review by armstrongsouljah — COMMENTED** [review]
Add the changes I left in the comments.

**Comment by ja-odur** [comment]
@malfahad please give more context on the `opt out` endpoint because I do get the same response each time I tried to switch between the options.

**Comment by malfahad** [comment]
@ja-odur  the opt out endpoint works this way;

`PUT /api/notifications/perm/<permission-type>`
 will opt you turn on the notification, calling this twice gives a message `You already opted into this notification` 
##### by default, all notifications are turned on.

`GET /api/notifications/perm/<token>` 
will turn off notifications. It is sent in the notificaion email.

**Comment by ja-odur** [comment]
> @ja-odur the opt out endpoint works this way;
> 
> `PUT /api/notifications/perm/<permission-type>`
> will opt you turn on the notification, calling this twice gives a message `You already opted into this notification`
> 
> ##### by default, all notifications are turned on.
> `GET /api/notifications/perm/<token>`
> will turn off notifications. It is sent in the notificaion email.

@malfahad  what is this token being passed? is it the same token as the `auth` token.

**Comment by malfahad** [comment]
@ja-odur the token generated is a jwt as well

File: authors/apps/core/models.py [hunk 14]
```diff
@@ -76,3 +73,51 @@ def send(self,email_text):
...
+
+class EmailNotificationDispatch:
+    def __init__(self):
+        self.sender_email = 'authors.haven.5@gmail.com'
+        self.subject = "AH - You Have a new notificaton"
+
+    def make_token(self, email, permission_type):
+        return jwt.encode({'email': email,
+                            'permission_type':permission_type,
+                             'exp': datetime.datetime.utcnow() + datetime.timedelta(days=30)
```
Comment (ja-odur) [inline authors/apps/core/models.py]: @malfahad set expiry date to 1day

**Review by MuhweziDeo — COMMENTED** [review]
Email Notification works okay

File: authors/apps/notifications/backends.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        """
+        notify users of new followers on their profiles
+
+        Args:
+            - request(Request): request object of the follower
+            - followed_profile(ProfileModel):
+
+        """
+        follower_profile = request.user.profile
+        if 'PROFILE_FOLLOW' not in followed_profile.notification_perms or \
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad for such scenarios, use a mapper such as a dictionary to prevent typos
```
{ 'PROFILE_FOLLOW': 'PROFILE_FOLLOW' }
```

File: authors/apps/notifications/backends.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+        self._dispatch(request, (follower_profile,
+                                 followed_profile, message,
+                                 action_link, ), 'PROFILE_FOLLOW')
+
+    def article_created(self, request, article):
+        """ 
+        notify author's followers about creation of new articles.
+
+        Args:
+            - request (Request); request object of the author
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad If I am right, the docstring convention does not have `-` prefix

**Comment by MuhweziDeo** [comment]
@malfahad  could u list all available parameters for notification types

File: authors/apps/notifications/backends.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+            - article (ArticleModel): The article object
+
+        """
+        people_to_notify = []
+        for person in article.author.followers.all():
+            if 'FOLLOWED_AUTHOR_ARTICLE_CREATED' in \
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad this can also be added to the mapper mentioned earlier

File: authors/apps/notifications/backends.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+                                        )
+        action_link = request.build_absolute_uri(reverse(
+            'comments:comment-detail', 
+            kwargs={'slug': comment.article.slug, 'pk': comment.pk}))
+        self._dispatch(request, (sender,
+                       receiver, message, action_link,),
+                       'AUTHORED_ARTICLE_COMMENTED')
+
+        for profile in comment.article.favorites.all():
+            if 'FAVORITED_ARTICLE_COMMENTED' in profile.notification_perms:
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad the same, use a mapper

File: authors/apps/notifications/backends.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+            - article(ArticleModel): the article object
+            - liked_by (ProfileModel): the liker's profile object, Defaults to None
+            - favorited_by (ProfileModel): the favoriter's profile object, Defaults to None  
+
+        """
+        sender = liked_by or favorited_by
+        receiver = article.author
+
+        if self._is_same_profile(sender, receiver):
+            return
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py:114]: @malfahad Is there a specific reason for this `falsey return`? I think raising an exception would be better.

File: authors/apps/notifications/backends.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        builds a notification message string
+        Args:
+            - person(ProfileModel): person mentioned in the notification
+            - message(String): message describing changes
+            - title(String): affected article title
+
+        returns:
+            - notification_message(String)
+        """
+        return self._get_name(person)\
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py:164]: @malfahad use format. it is much cleaner than concatenating strings using the `+` operator.

File: authors/apps/notifications/backends.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+        sender, receiver, message, action_link = notification
+        if permission_type in receiver.notification_perms:
+            notification = Notification.objects.create_notification(
+                sender=sender,
+                receiver=receiver,
+                message=message,
+                action_link=action_link
+                )
+  
+        if 'RECEIVE_NOTIFICATION_EMAILS' in receiver.notification_perms:
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad use a mapper here as well

File: authors/apps/notifications/backends.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+                                            permission_type])
+            thread.setDaemon(True)
+            thread.start()
+
+    def _get_name(self, sender):
+        """
+        Builds a person's name from a ProfileModel object.
+        
+        Args:
+            - sender(ProfileModel): the sender's object
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad follow convention

File: authors/apps/notifications/backends.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+        
+        returns:
+            -name(String); sender's full name or username
+
+        """
+
+        return sender.first_name + " " + sender.last_name \
```
Comment (ja-odur) [inline authors/apps/notifications/backends.py]: @malfahad use format here as well

File: authors/apps/notifications/models.py [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+
+class NotificationManager(models.Manager):
+    
+    def create_notification(self, sender, receiver, message, action_link=None):
+        """Create and return a  notification"""
+        if sender not in Profile.objects.all():
+            raise TypeError('Notification must have a valid sender.')
+
+        if receiver not in Profile.objects.all():
+            raise TypeError('Notification must have a valid receiver')
```
Comment (ja-odur) [inline authors/apps/notifications/models.py]: @malfahad the message is similar to that on line `9`. this can be extracted and carefully using `format` to fill in the different ending words.

File: authors/apps/notifications/views.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+class NotiificationsAPIView(ListAPIView):
+    permission_classes = (AllowAny,)
+    renderer_classes = (NotificationJSONRenderer,)
+    serializer_class = NotificationSerializer
+
+    def get_queryset(self):
+        user = self.request.user.profile
+        task = self.kwargs.get('task')
+        if task.lower() == 'all':
```
Comment (ja-odur) [inline authors/apps/notifications/views.py]: @malfahad create a separate mapper for this

File: authors/apps/notifications/backends.py (hunk 19 above)
Comment (malfahad) [reply authors/apps/notifications/backends.py:114]: Yes, for example, comments, The notification creation is dispatched when a comment is saved. Take for example creating a comment on your own article. This falsey return stops the app from notifying one about their own actions.

File: authors/apps/notifications/utils.py [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    MY_ARTICLE_COMMENTED = 'MY_ARTICLE_COMMENTED'
+    FAVORITED_ARTICLE_COMMENTED = 'FAVORITED_ARTICLE_COMMENTED'
+    MY_ARTICLE_LIKED = 'MY_ARTICLE_LIKED'
+    FAVORITED_ARTICLE_LIKED = 'FAVORITED_ARTICLE_LIKED'
+    AUTHORED_ARTICLE_COMMENTED = 'AUTHORED_ARTICLE_COMMENTED'
+    AUTHORED_ARTICLE_LIKED = 'AUTHORED_ARTICLE_LIKED'
+    AUTHORED_ARTICLE_FAVORITED = 'AUTHORED_ARTICLE_FAVORITED'
+ 
+    all = {
+        'RECEIVE_NOTIFICATION_EMAILS': 'RECEIVE_NOTIFICATION_EMAILS',
```
Comment (ja-odur) [inline authors/apps/notifications/utils.py]: @malfahad you can now refer to the class instances here inorder to have one single place with a hardcoded string e.g
```suggestion
        'RECEIVE_NOTIFICATION_EMAILS': cls.RECEIVE_NOTIFICATION_EMAILS,
```

**Review by MuhweziDeo — APPROVED** [review]

**Review by d-rita — APPROVED** [review]
LGTM

**Review by ja-odur — APPROVED** [review]
LGTM

**Review by armstrongsouljah — APPROVED** [review]

**Review by kdaglas — APPROVED** [review]
LGTM nice work Fahad

**Review by calvinpete — APPROVED** [review]
LGTM
